catalog/dags/data_refresh/create_filtered_index.py

```diff
@@ -0,0 +1,133 @@
+"""
+# Create filtered index TaskGroup factory
+
+This module contains factory functions used to generate the Airflow tasks for
+creating filtered indices, used by the data refreshes of each media type and the
+Create Filtered Index DAGs.
+
+Filtered index creation is handled by the ingestion server. The TaskGroups generated
+by the ``create_filtered_index_creation_task_groups`` function in this module are
+responsible for creating and promoting filtered indices. The ``create_filtered_index``
+TaskGroup trigges the ingestion server action to create and populate the filtered
+index for a given media type, and awaits the completion of the index creation. The
+``promote_filtered_index`` TaskGroup awaits healthy results from the newly created
+index, and then points the filtered index alias for the media type to the new index,
+finally deleting the old, now unused filtered index. These TaskGroups are used in
+the data refresh DAGs to execute the filtered index steps.
+"""
+from datetime import timedelta
+
+from airflow.operators.empty import EmptyOperator
+from airflow.utils.task_group import TaskGroup
+from airflow.utils.trigger_rule import TriggerRule
+
+from common import ingestion_server
+from common.constants import XCOM_PULL_TEMPLATE
+from data_refresh.data_refresh_types import DataRefresh
+
+
+def create_filtered_index_creation_task_groups(
+    data_refresh: DataRefresh,
+    origin_index_suffix: str | None,
+    destination_index_suffix: str | None,
+) -> tuple[TaskGroup, TaskGroup]:
+    """
+    Create the TaskGroups that performs filtered index creation and promotion for
+    the given DataRefresh.
+    """
+    media_type = data_refresh.media_type
+    target_alias = f"{media_type}-filtered"
+
+    def create_and_populate_filtered_index(
```

Comment:
I know these functions were within the DAG function in the previous implementation, but it does feel a little odd to have them be a pure function within a pure function  in the new setup. Do you think it makes sense to try and extract these out so they can be functions at the top level that then get called within `create_filtered_index_creation_task_groups` rather than defined within it? Or is it easier to keep as-is?

Reply:
We certainly can, it just means we have to pass more params through to those methods.